src/manage_nsfs/nc_lifecycle.js

```diff
@@ -472,22 +421,341 @@ class NCLifecycle {
+            const mmapply_policy_res = await os_utils.exec(`mmapplypolicy ${mount_point_path} -P ${ilm_policy_tmp_path} -f ${ILM_CANDIDATES_TMP_DIR} -I defer`, { return_stdout: true });
+            dbg.log2('create_candidates_file_by_gpfs_ilm_policy mmapplypolicy res ', mmapply_policy_res);
+        } catch (err) {
+            throw new Error(`create_candidates_file_by_gpfs_ilm_policy failed with error ${err}`);
+        }
+    }
+
+    /**
+     * parse_candidates_from_gpfs_ilm_policy
+     * TODO - when adding resume, we should read the key_marker from current status or from the rule log file
+     * TODO - move dbg0 to dbg2
+     * @param {Object} bucket_json
+     * @param {*} lifecycle_rule
+     * @param {String} rule_candidates_path 
+     * @returns 
+     */
+    async parse_candidates_from_gpfs_ilm_policy(bucket_json, lifecycle_rule, rule_candidates_path) {
+        const parsed_candidates_array = [];
+        let reader;
+        try {
+            let rule_state = this.lifecycle_run_status.buckets_statuses[bucket_json.name].rules_statuses[lifecycle_rule.id]?.state || {};
+            const ilm_candidates_file_exists = await native_fs_utils.stat_ignore_enoent(this.non_gpfs_fs_context, rule_candidates_path);
+            if (!ilm_candidates_file_exists) {
+                dbg.log2(`parse_candidates_from_gpfs_ilm_policy ilm_candidates_file_exists does not exist, no candidates to delete`);
+                rule_state = { candidates_file_offset: undefined, read_lines: undefined, is_finished: true };
+                this.lifecycle_run_status.buckets_statuses[bucket_json.name].rules_statuses[lifecycle_rule.id].state = rule_state;
+                return [];
+            }
+            dbg.log2(`parse_candidates_from_gpfs_ilm_policy existing rule_state=${util.inspect(rule_state)}`);
+            // set offset so we won't need to read the file from the beginning
+            reader = new NewlineReader(this.non_gpfs_fs_context, rule_candidates_path, { lock: 'SHARED', read_file_offset: rule_state?.candidates_file_offset || 0 });
+            const [count, is_finished] = await reader.forEachFilePathEntry(async entry => {
+                dbg.log2(`parse_candidates_from_gpfs_ilm_policy parsed_candidates_array=${util.inspect(parsed_candidates_array)} NC_LIFECYCLE_LIST_BATCH_SIZE=${config.NC_LIFECYCLE_LIST_BATCH_SIZE}`);
+                //if (parsed_candidates_array.length >= config.NC_LIFECYCLE_LIST_BATCH_SIZE) {
+                if (parsed_candidates_array.length >= config.NC_LIFECYCLE_LIST_BATCH_SIZE) {
+                    return false; // stop reading
+                }
+                rule_state = {
+                    candidates_file_offset: reader.next_line_file_offset, is_finished: false
+                };
+                dbg.log2('parse_candidates_from_gpfs_ilm_policy: reader.read_file_offset', reader.read_file_offset, 'reader.next_line_file_offset: ', reader.next_line_file_offset);
```

Comment:
can you combine logs. and if there are, remove redundant logs. this function is hard to understand as it is